pennylane/templates/state_preparations/mottonen.py

```diff
@@ -128,102 +140,86 @@ def _uniform_rotation_dagger(gate, alpha, control_wires, target_wire):
+    return np.sum(diff, axis=1)
 
 
 def _get_alpha_y(a, n, k):
-    r"""Computes the rotation angles alpha for the Y rotations.
+    r"""Computes the rotation angles required to implement the uniformly-controlled Z rotation
+    applied to the :math:`k`th qubit.
+
+    The :math:`j`th angle is related to the absolute values a of the desired amplitudes via:
+
+    .. math:: \alpha^{y,k}_j = 2 \arcsin \sqrt{ \frac{ \sum_{l=1}^{2^{k-1}} a_{(2j-1)2^{k-1} +l}^2  }{ \sum_{l=1}^{2^{k}} a_{(j-1)2^{k} +l}^2  } }
 
     Args:
-        omega (float): phase of the input
-        n (int): total number of qubits
+        a (array): absolute values of the state to prepare
+        n (int): total number of qubits for the uniformly-controlled rotation
         k (int): index of current qubit
 
     Returns:
-        scipy.sparse.dok_matrix[np.float64]: a sparse vector representing :math:`\alpha^y_k`
+        array representing :math:`\alpha^{y,k}`
     """
+    indices_numerator = [
+        [(2 * (j + 1) - 1) * 2 ** (k - 1) + l for l in range(2 ** (k - 1))]
+        for j in range(2 ** (n - k))
+    ]
+    numerator = np.take(a, indices=indices_numerator)
+    numerator = np.sum(np.abs(numerator) ** 2, axis=1)
 
-    alpha = sparse.dok_matrix((2 ** (n - k), 1), dtype=np.float64)
-
-    numerator = sparse.dok_matrix((2 ** (n - k), 1), dtype=np.float64)
-    denominator = sparse.dok_matrix((2 ** (n - k), 1), dtype=np.float64)
-
-    for (i, _), e in a.items():
-        j = int(math.ceil((i + 1) / 2 ** k))
-        l = (i + 1) - (2 * j - 1) * 2 ** (k - 1)
-        is_part_numerator = 1 <= l <= 2 ** (k - 1)
-
-        if is_part_numerator:
-            numerator[j - 1, 0] += e * e
-        denominator[j - 1, 0] += e * e
-
-    for (j, _), e in numerator.items():
-        numerator[j, 0] = math.sqrt(e)
-    for (j, _), e in denominator.items():
-        denominator[j, 0] = 1 / math.sqrt(e)
+    indices_denominator = [
+        [j * 2 ** k + l for l in range(0, int(2 ** k))] for j in range(2 ** (n - k))
+    ]
+    denominator = np.take(a, indices=indices_denominator)
+    denominator = np.sum(np.abs(denominator) ** 2, axis=1)
 
-    pre_alpha = numerator.multiply(denominator)  # type: sparse.csr_matrix
-    for (j, _), e in pre_alpha.todok().items():
-        alpha[j, 0] = 2 * np.arcsin(e)
+    # Divide only where denominator is zero, else leave initial value of zero.
+    # The equation guarantees that the numerator is also zero in the corresponding entries.
+    division = np.divide(
+        numerator, denominator, out=np.zeros_like(numerator, dtype=float), where=denominator != 0.0
+    )
 
-    return alpha
+    return 2 * np.arcsin(np.sqrt(division))
 
 
 @template
 def MottonenStatePreparation(state_vector, wires):
     r"""
     Prepares an arbitrary state on the given wires using a decomposition into gates developed
-    by Möttönen et al. (Quantum Info. Comput., 2005).
+    by `Möttönen et al. (2004) <https://arxiv.org/pdf/quant-ph/0407010.pdf>`_.
```

Comment:
Add umlauts to "Mottonen" below in line 227 (I can't seem to add a comment directly to that line during the review)